Mathlib/Analysis/PSeries.lean

```diff
@@ -19,84 +19,137 @@ converges if and only if so does `∑ k, 2 ^ k f (2 ^ k)`. We prove this test in
-  have : ∀ k ∈ Ico (2 ^ n + 1) (2 ^ (n + 1) + 1), f (2 ^ (n + 1)) ≤ f k := by
-    -- Note(kmill): was `fun k hk => ...` but `mem_Ico.mp hk` was elaborating with some
-    -- delayed assignment metavariables that weren't resolved in time. `intro` fixes this.
-    intro k hk
-    exact hf (Nat.one_le_two_pow.trans_lt <| (Nat.lt_succ_of_le le_rfl).trans_le (mem_Ico.mp hk).1)
-      (Nat.le_of_lt_succ <| (mem_Ico.mp hk).2)
+      (add_le_add_right (hu n.zero_le) _ : u 0 + 1 ≤ u n + 1),
+      add_le_add_right (hu n.le_succ) _]
+  have : ∀ k ∈ Ico (u n + 1) (u (n + 1) + 1), f (u (n + 1)) ≤ f k := fun k hk =>
+    hf (Nat.lt_of_le_of_lt (Nat.succ_le_of_lt (h_pos n)) <| (Nat.lt_succ_of_le le_rfl).trans_le
+      (mem_Ico.mp hk).1) (Nat.le_of_lt_succ <| (mem_Ico.mp hk).2)
   convert sum_le_sum this
   simp [pow_succ, mul_two]
+
+theorem sum_condensed_le' (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) (n : ℕ) :
+    (∑ k in range n, 2 ^ k • f (2 ^ (k + 1))) ≤ ∑ k in Ico 2 (2 ^ n + 1), f k := by
+  convert sum_schlomilch_le' hf (fun n => pow_pos zero_lt_two n)
+    (fun m n hm => pow_le_pow_right one_le_two hm) n using 2
+  simp [pow_succ, mul_two, two_mul]
 #align finset.sum_condensed_le' Finset.sum_condensed_le'
 
+theorem sum_schlomilch_le {C : ℕ} (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (h_nonneg : ∀ n, 0 ≤ f n) (hu : Monotone u) (h_succ_diff : SuccDiffBounded C u) (n : ℕ) :
+    ∑ k in range (n + 1), (u (k + 1) - u k) • f (u k) ≤
+    (u 1 - u 0) • f (u 0) + C • ∑ k in Ico (u 0 + 1) (u n + 1), f k := by
+  rw [sum_range_succ', add_comm]
+  gcongr
+  suffices ∑ k in range n, (u (k + 2) - u (k + 1)) • f (u (k + 1)) ≤
+  C • ∑ k in range n, ((u (k + 1) - u k) • f (u (k + 1))) by
+    refine this.trans (nsmul_le_nsmul_right ?_ _)
+    exact sum_schlomilch_le' hf h_pos hu n
+  have : ∀ k ∈ range n, (u (k + 2) - u (k + 1)) • f (u (k + 1)) ≤
+    C • ((u (k + 1) - u k) • f (u (k + 1))) := by
+    intro k _
+    rw [smul_smul]
+    gcongr
+    · exact h_nonneg (u (k + 1))
+    exact mod_cast h_succ_diff k
+  convert sum_le_sum this
+  simp [smul_sum]
+
 theorem sum_condensed_le (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) (n : ℕ) :
     (∑ k in range (n + 1), 2 ^ k • f (2 ^ k)) ≤ f 1 + 2 • ∑ k in Ico 2 (2 ^ n + 1), f k := by
   convert add_le_add_left (nsmul_le_nsmul_right (sum_condensed_le' hf n) 2) (f 1)
   simp [sum_range_succ', add_comm, pow_succ', mul_nsmul', sum_nsmul]
 #align finset.sum_condensed_le Finset.sum_condensed_le
-
 end Finset
 
 namespace ENNReal
 
-open Filter BigOperators
+open Filter BigOperators Finset
 
-variable {f : ℕ → ℝ≥0∞}
+variable {u : ℕ → ℕ} {f : ℕ → ℝ≥0∞}
+
+open NNReal in
+theorem le_tsum_schlomilch (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (hu_strict : StrictMono u) :
+    ∑' k , f k ≤ ∑ k in range (u 0), f k + ∑' k : ℕ, (u (k + 1) - u k) * f (u k) := by
+  have hu : Monotone u := by
+    apply StrictMono.monotone hu_strict
```

Comment:
I'd just use `hu_strict.monotone` in the one place you need it below. Also, maybe rename `hu_strict` to `hu`